hiero-observability/hiero-metrics-core/src/main/java/org/hiero/metrics/api/core/MetricRegistry.java

```diff
@@ -0,0 +1,204 @@
+     * @return the registered metric, never {@code null}
+     * @throws IllegalArgumentException if a metric with the same name already exists in the registry
+     */
+    @NonNull
+    <M extends Metric, B extends Metric.Builder<?, M>> M register(@NonNull B builder);
+
+    /**
+     * Finds a metric by its key.
+     *
+     * @param key the metric key, must not be {@code null}
+     * @param <M> the type of the metric
+     * @return an {@link Optional} for the found metric. If no metric is found, an empty {@link Optional} is returned.
+     */
+    @NonNull
+    <M extends Metric> Optional<M> findMetric(@NonNull MetricKey<M> key);
+
+    /**
+     * Gets a metric by its key.
+     *
+     * @param key the metric key, must not be {@code null}
+     * @param <M> the type of the metric
+     * @return the found metric, never {@code null}
+     * @throws IllegalArgumentException if no metric is found for the given key
+     */
+    @NonNull
+    default <M extends Metric> M getMetric(@NonNull MetricKey<M> key) {
+        Optional<M> metric = findMetric(key);
+        if (metric.isPresent()) {
+            return metric.get();
+        }
+        throw new IllegalArgumentException("Metric not found: " + key);
+    }
+
+    /**
+     * Resets all registered metrics to their initial state.
+     * Default implementation iterates over all metrics and calls their {@link Metric#reset()} method.
+     */
+    default void reset() {
+        metrics().forEach(Metric::reset);
+    }
+
+    /**
+     * Creates a new {@link Builder} for constructing a {@link MetricRegistry}.
+     *
+     * @param registryName the name of the registry, must not be {@code null} or blank
+     * @return a new {@link Builder} instance
+     */
+    @NonNull
+    static Builder builder(@NonNull String registryName) {
+        return new Builder(registryName);
+    }
+
+    /**
+     * Builder for creating instances of {@link MetricRegistry}.
+     */
+    class Builder {
+
+        private static final Logger logger = LogManager.getLogger(MetricRegistry.class);
+
+        private final String registryName;
+        private boolean discoreMetricProviders = false;
+        private final List<Label> globalLabels = new ArrayList<>();
+        private final Set<String> globalLabelNames = new HashSet<>();
+
+        public Builder(@NonNull String registryName) {
+            this.registryName = ArgumentUtils.throwArgBlank(registryName, "registry name");
+        }
+
+        /**
+         * Adds a global label to the registry that will be applied to all metrics.
+         *
+         * @param label the label to add, must not be {@code null}
+         * @return this builder instance
+         * @throws NullPointerException if the label is {@code null}
+         * @throws IllegalArgumentException if a global label with the same name already exists
+         */
+        @NonNull
+        public Builder addGlobalLabel(@NonNull Label label) {
```

Comment:
To be consistent with metric builders, these methods should be named `withXYZ()`, not `addXYZ()`